Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp

```diff
@@ -281,6 +281,27 @@ void GraphicsContextCG::restore(GraphicsContextState::Purpose purpose)
     m_userToDeviceTransformKnownToBeIdentity = false;
 }
 
+#if HAVE(SUPPORT_HDR_DISPLAY_APIS)
+static void setCGDynamicRangeLimitForImage(CGContextRef context, CGImageRef image, float dynamicRangeLimit)
+{
+    float edrStrength = dynamicRangeLimit == 1.0 ? 1 : 0;
+    float cdrStrength = dynamicRangeLimit == 0.5 ? 1 : 0;
+    unsigned averageLightLevel = CGImageGetContentAverageLightLevelNits(image);
+
+    RetainPtr edrStrengthNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberFloatType, &edrStrength));
+    RetainPtr cdrStrengthNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberFloatType, &cdrStrength));
+    RetainPtr averageLightLevelNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &averageLightLevel));
+
+    CFTypeRef toneMappingKeys[] = { kCGContentEDRStrength, kCGContentAverageLightLevel, kCGConstrainedDynamicRange };
+    CFTypeRef toneMappingValues[] = { edrStrengthNumber.get(), averageLightLevelNumber.get(), cdrStrengthNumber.get() };
+
+    RetainPtr toneMappingOptions = adoptCF(CFDictionaryCreate(kCFAllocatorDefault, toneMappingKeys, toneMappingValues, sizeof(toneMappingKeys) / sizeof(toneMappingKeys[0]), &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
+
+    CGContentToneMappingInfo toneMappingInfo = { kCGToneMappingReferenceWhiteBased, toneMappingOptions.get() };
+    CGContextSetContentToneMappingInfo(context, toneMappingInfo);
```

Comment:
Is this something that is part of the CGContexts gstate and therefore part of the save/restore system?

Reply:
Yes. And this is why at the end of this function we do:
```
    if (!stateSaver.didSave()) {
        ...    
        CGContextSetContentToneMappingInfo(context, oldToneMappingInfo);
        ...
    }
```

Reply:
Yes this is correct. If we add

```
GraphicsContext::setEDRTargetHeadroom()
GraphicsContext::setContentToneMappingInfo()
```

Then we will have to add the `EDRTargetHeadroom` and `ContentToneMappingInfo` through our own `GraphicsContextState`